File: backend/src/utils/embedding.ts
import crypto from "node:crypto";

const VECTOR_DIMENSION = 768;

export function fallbackEmbeddingFromText(text: string): number[] {
  const hash = crypto.createHash("sha256").update(text).digest();
  const vector: number[] = [];
  for (let i = 0; i < VECTOR_DIMENSION; i += 1) {
    const byte = hash[i % hash.length];
    const normalized = (byte / 255) * 2 - 1; // [-1, 1]
    vector.push(Number(normalized.toFixed(6)));
  }
  return vector;
}

File: rag-service/app/core/embedding.py
import hashlib

VECTOR_DIMENSION = 768


def fallback_embedding(text: str) -> list[float]:
    digest = hashlib.sha256(text.encode("utf-8")).digest()
    values: list[float] = []
    for index in range(VECTOR_DIMENSION):
        byte = digest[index % len(digest)]
        normalized = (byte / 255) * 2 - 1
        values.append(round(normalized, 6))
    return values
